File: Ruby/remove_nth_from_end.rb
class ListNode
    attr_accessor :val, :next
    def initialize(val = 0, _next = nil)
        @val = val
        @next = _next
    end
end

def print(list)
  curr = list
  loop do
    p curr.val
    break if curr.next.nil?
    curr = curr.next
  end
end


# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
    return nil if head.nil?

    size = 0
    curr = head
    loop do
      break if curr.nil?
      size += 1
      curr = curr.next
    end

    nth = size - n

    # if nth <= 0
    #   head = head.next
    #   return head
    # end
    return head&.next if nth <= 0

    p "SIZE: #{size}, NTH: #{nth}"

    m = 0
    curr = head
    loop do
      m += 1
      if nth == m
        curr.next = curr&.next&.next
        break
      end
      curr = curr.next
    end

    return head
end

# head = ListNode.new(1, ListNode.new(2, ListNode.new(3, ListNode.new(4, ListNode.new(5)))))
head = ListNode.new(1, ListNode.new(2))

head = remove_nth_from_end(head, 2)

print(head)

File: Java/RemoveNthFromEnd.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RemoveNthFromEnd {
    class ListNode {
        public int data;
        public ListNode next;

        public ListNode() {
            data = 0;
            next = null;
        }

        public ListNode(int data, ListNode next) {
            this.data = data;
            this.next = next;
        }
    }

    public void print(ListNode node) {
        ListNode curr = node;
        while (curr != null) {
            System.out.println(curr.data);
            curr = curr.next;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null) return null;
        int size = 0;
        Map<Integer, ListNode> nodeMap = new HashMap<>();

        ListNode curr = head;
        while (curr != null) {
            size++;
            nodeMap.put(size, curr);
            curr = curr.next;
        }

//        System.out.println("SIZE: " + size);

        int nth = size - n;
//        System.out.println("NTH: " + nth);

        if (nth == 0) {
            head = head.next;
//            print(head);
            return head;
        }

        if (nth == (size - 1)) {
            nodeMap.get(nth).next = null;
        } else {
            nodeMap.get(nth).next = nodeMap.get(nth + 2);
        }

//        print(head);
        return head;

    }

    public void func() {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, null)))));
        removeNthFromEnd(head, 5);
    }

    public static void main (String[] args) {
        RemoveNthFromEnd obj = new RemoveNthFromEnd();
        obj.func();
    }
}
